### PR TITLE
Dut monitor

### PR DESCRIPTION
**Pull request by yvolynets-mlnx**
### Description of PR

Summary: Added HW resources monitoring feature  described in [HLD](https://github.com/yvolynets-mlnx/SONiC/blob/dut_monitor/doc/DUT_monitor_HLD.md)
Fixes # (issue)

### Type of change

- [] Bug fix
- [x] Testbed and Framework(new/improvement)
- [] Test case(new/improvement)

#### How did you do it?
Developed new modules located in tests/plugins/

#### How did you verify/test it?
Tested on platform tests on local setup.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -33,17 +33,17 @@ def __init__(self, testbed_file):
             topo = csv.DictReader(f)
             for line in topo:
                 tb_prop = {}
-                name = ''
+                name = ""
```
Comment (stepanblyschak) [inline tests/conftest.py]: What is the purpose of the change?

File: tests/conftest.py [hunk 2]
```diff
@@ -67,6 +67,8 @@ def testbed(request):
         raise ValueError("testbed and testbed_file are required!")
 
     tbinfo = TestbedInfo(tbfile)
+    tb_topo = tbinfo.testbed_topo[tbname]
```
Comment (stepanblyschak) [inline tests/conftest.py]: Where is this used?

File: tests/plugins/pytest_dut_monitor.py [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+
+
+def pytest_unconfigure(config):
+    dut_monitor = getattr(config, "dut_monitor", None)
+    if dut_monitor:
+        del config.dut_monitor
+        config.pluginmanager.unregister(dut_monitor)
+
+
+class HDDMonitorError(Exception):
```
Comment (stepanblyschak) [inline tests/plugins/pytest_dut_monitor.py]: I would like to suggest a structure like this:
```
tests
  plugins
    dut_monitor
        pytest_dut_monitor.py
        dut_monitor.py
```
And suggest to put *Error in errors.py in tests/plugins/dut_monitor/

File: tests/plugins/thresholds.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# Maybe use - CMD_PLATFORM_SUMMARY = "show platform summary"
+arc-switch1029:
```
Comment (stepanblyschak) [inline tests/plugins/thresholds.yml]: Why per testbed host name and not per platform/HWSKU ? Where are the default values?

File: tests/plugins/dut_monitor.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    result = general_template.format(timestamp=datetime.now().strftime("%Y-%m-%d %H:%M:%S"), total=total) + "\n".join(top_consumer_list)
+    log_file.write(result)
+
+
+def process_ram(log_file):
+    """
+    @summary: Fetch RAM utilization and write it to the file. Use 'MemTotal' and 'MemAvailable' from '/proc/meminfo' to obtain used RAM amount.
+    @param log_file: Opened file object to store fetched CPU utilization.
+    """
+    with open('/proc/meminfo') as file:
```
Comment (stepanblyschak) [inline tests/plugins/dut_monitor.py]: Don't use python built-in type 'file' as variable name

File: tests/plugins/dut_monitor.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    used_percent = used * 100 / total_mem_in_kb
+    log_file.write("\"{date}\": {used_ram}\n".format(date=datetime.now().strftime("%Y-%m-%d %H:%M:%S"), used_ram=used_percent))
+
+
+def process_hdd(log_file):
+    """
+    @summary: Fetch used amount of HDD and write it to the file.
+    @param log_file: Opened file object to store fetched CPU utilization.
+    """
+    hdd_usage = os.popen(FETCH_HDD_CMD).read().split("\n")[1].split()[4].rstrip("%")
```
Comment (stepanblyschak) [inline tests/plugins/dut_monitor.py]: What is 1 and what is 4?

File: tests/plugins/pytest_dut_monitor.py [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+
+from collections import OrderedDict
+from  datetime import datetime
+
+logger = logging.getLogger(__name__)
+DUT_MONITOR = "/tmp/dut_monitor.py"
+DUT_CPU_LOG = "/tmp/cpu.log"
+DUT_RAM_LOG = "/tmp/ram.log"
+DUT_HDD_LOG = "/tmp/hdd.log"
+THRESHOLDS = os.path.join(os.path.split(__file__)[0], "thresholds.yml")
```
Comment (stepanblyschak) [inline tests/plugins/pytest_dut_monitor.py]: Better to let user specify different threasholds file, what do you think? It's kind of change to let users of tests disintegrate their topology/DUT configurations from test logic

**Review by stepanblyschak — CHANGES_REQUESTED** [review]

File: tests/plugins/thresholds.yml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  cpu_total: 90
```
Comment (stepanblyschak) [inline tests/plugins/thresholds.yml]: BTW, how a single test can provide its own values? Any performance test/reboot test will utilize CPU at 100% causing such check to fail

File: tests/conftest.py (hunk 1 above)
Comment (yvolynets-mlnx) [reply tests/conftest.py]: To have all quotes in the current file in one style.

File: tests/conftest.py (hunk 2 above)
Comment (yvolynets-mlnx) [reply tests/conftest.py]: Needs to be fixed. Thank you.

File: tests/plugins/dut_monitor.py (hunk 5 above)
Comment (yvolynets-mlnx) [reply tests/plugins/dut_monitor.py]: Fixed. Thank you.

File: tests/conftest.py (hunk 1 above)
Comment (stepanblyschak) [reply tests/conftest.py]: IMO, we can't control any other guy who decides to use single quotes unless there is a code convention document and code convention checks tools integrated in CI
On the other hand such changes bloat the commit, introducing not related diff in history, e.g. person looking for last author of the line via 'git blame' will see you however there is no change in logic, code structure or functionality made by you.
I would suggest creating another PR dedicated for such cosmetic change so that others who browse the history of the file can effectively skip them, however that means you need to convince community to use this style and not the other which is not possible without agreed code conventions

File: tests/plugins/dut_monitor.py (hunk 6 above)
Comment (yvolynets-mlnx) [reply tests/plugins/dut_monitor.py]: Fixed. Thank you.

File: tests/conftest.py (hunk 1 above)
Comment (yvolynets-mlnx) [reply tests/conftest.py]: "I would suggest creating another PR dedicated for such cosmetic change so that others who browse the history of the file can effectively skip them" - agree.
"however that means you need to convince community to use this style and not the other which is not possible without agreed code conventions" - project should have agreed style convention.

So lets leave this as it is?

Comment (stepanblyschak) [reply tests/conftest.py]: "So lets leave this as it is?" contradicts with you agreement on another PR, my preference is another PR that fixes this in all pytest tests.
" project should have agreed style convention" - ideally yes, sonic doesn't

Comment (yvolynets-mlnx) [reply tests/conftest.py]: Ok. This minor updates will be removed. Thank you.

File: tests/plugins/pytest_dut_monitor.py [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+        """
+        self.running = True
+        self._upload_to_dut()
+        logger.debug("Start HW resuorces monitoring on the DUT...")
+
+        self.run_channel = self.ssh.get_transport().open_session()
+        self.run_channel.get_pty()
+        self.run_channel.settimeout(5)
+        # Start monitoring on DUT
+        self.run_channel.exec_command("python {} --start".format(DUT_MONITOR))
```
Comment (stepanblyschak) [inline tests/plugins/pytest_dut_monitor.py]: How this survives when test performs reboot?

File: tests/plugins/pytest_dut_monitor.py (hunk 3 above)
Comment (yvolynets-mlnx) [reply tests/plugins/pytest_dut_monitor.py]: Fixed. Thank you.

File: tests/plugins/pytest_dut_monitor.py (hunk 9 above)
Comment (yvolynets-mlnx) [reply tests/plugins/pytest_dut_monitor.py]: There is mechanism to track network connectivity and restore monitoring script if network connectivity dropped.
It starts here:
self._thread = threading.Thread(name="Connection tracker", target=self._track_connection)
self._thread.setDaemon(True)
self._thread.start()

File: tests/plugins/thresholds.yml (hunk 4 above)
Comment (yvolynets-mlnx) [reply tests/plugins/thresholds.yml]: Fixed.

File: tests/plugins/pytest_dut_monitor.py (hunk 7 above)
Comment (yvolynets-mlnx) [reply tests/plugins/pytest_dut_monitor.py]: Added possibility to specify thresholds file via "--thresholds_file" pytest option.

File: tests/plugins/thresholds.yml (hunk 8 above)
Comment (yvolynets-mlnx) [reply tests/plugins/thresholds.yml]: Fixed. Thresholds can be updated by any test case. For this in test case  user should pass "dut_monitor" fixture.

In test case thresholds can be updated like:
dut_monitor["cpu_total"] = 70
dut_monitor["ram_peak"] = 50
...
Full list of properties are defined in default thresholds file.
